architecture/2025-04-compute-extensibility.md

```diff
@@ -0,0 +1,474 @@
+# Compute Platform Extensibility
+
+* **Author**: Brooke Hamilton (@Brooke-Hamilton)
+
+## Summary
+
+Radius will provide extensible support for multiple compute platforms through recipes rather than hard-coded support for each platform.
+
+Core resource types (`containers`, `gateways`, and `secretStores`) will allow recipes to be registered for them, with default recipes for Kubernetes and Azure Container Instances (ACI) provided by Radius. Customers can use, modify, or remove the default recipes.
+
+This design enables:
+
+* Architectural separation of Radius core logic from platform provisioning code
+* Community-provided extensions to support new compute platforms without Radius code changes
+* Consistent platform engineering experience across all resource types
+
+We have two options for implementation:
+
+1. (Recommended) Create UDT versions of the core resource types (`containers`, `gateways`, and `secretStores`) paired with recipe-based provisioning for ACI and Kubernetes. Later, remove the built-in core types and existing Kubernetes provisioning code. Only `environments` and `applications` will remain as built-in core types.
+2. Add recipe support to the built-in core types and create provisioning recipes for ACI and Kubernetes. Later, remove the hard-coded provisioning code.
+
+The primary criterion for deciding between option 1 and 2 is whether we want the application model to be represented as Radius-provided UDTs, or if we want to continue to have the core resource types that represent the application model to be built-in.
+
+Both options have technical risk, primarily in the implementation of Kubernetes provisioning with recipes and the associated separation of provisioning code from the Radius application. Option 2 has lower initial risk and faster time to initial release because the scope of the initial release is smaller.
+
+## Terms and definitions
+
+| Term | Definition |
+|------|------------|
+| **Compute platform** | An environment where applications can be deployed and run, such as Kubernetes, Azure Container Instances (ACI), etc. |
+| **Core types** | Built-in resource types provided by Radius, including `containers`, `gateways`, `secretStores`, `environments`, and `applications`. |
+| **User-defined type (UDT)** | A custom resource type defined separately from Radius core types and loaded into Radius without requiring a new Radius release. |
+| **Recipe** | A set of instructions that provisions a Radius resource to a Radius environment. Recipes are implemented in Bicep or Terraform. |
+| **Resource Provider (RP)** | A component responsible for handling create, read, update, delete, and list (CRUDL) operations for a specific resource type. |
+
+### Goals
+
+* Provide platform engineers with the ability to deploy to specific platforms other than Radius, like ACI.
+* Provide a recipe-based platform engineering experience that is consistent for user-defined types and core types.
```

Comment:
We need a step to validate the recipe that is being registered to not have runtime errors.

Reply:
We are tracking recipe validation in the development plan.